Change the linebreak behavior of minlength and maxlength attributes of TEXTAREA

cdumez opened the pull request:
#### 2252898e2468994720be25176170e00bc7b26ca3
<pre>
Change the linebreak behavior of minlength and maxlength attributes of TEXTAREA
<a href="https://bugs.webkit.org/show_bug.cgi?id=249916">https://bugs.webkit.org/show_bug.cgi?id=249916</a>
rdar://103856355

Reviewed by Darin Adler.

Align the linebreak behavior of minlength and maxlength attributes of TEXTAREA with
the specification, Blink and Gecko.

Old behavior:
 CRLF, CR, or LF is counted as 2 because we limited the submission value.

New behavior:
 CRLF, CR, or LF is counted as 1 because we limit the API value.

This is a cherry-pick of Blink:
<a href="https://chromium.googlesource.com/chromium/src.git/+/a15474ff7adff41a8956ab436ac63a7d5b7a090f%5E%21/#F3">https://chromium.googlesource.com/chromium/src.git/+/a15474ff7adff41a8956ab436ac63a7d5b7a090f%5E%21/#F3</a>

* LayoutTests/fast/forms/textarea-maxlength-expected.txt:
* LayoutTests/fast/forms/textarea-maxlength.html:
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::computeLengthForAPIValue):
(WebCore::HTMLTextAreaElement::handleBeforeTextInsertedEvent const):
(WebCore::HTMLTextAreaElement::sanitizeUserInputValue):
(WebCore::HTMLTextAreaElement::validationMessage const):
(WebCore::HTMLTextAreaElement::tooShort const):
(WebCore::HTMLTextAreaElement::tooLong const):
(WebCore::computeLengthForSubmission): Deleted.
(WebCore::numberOfLineBreaks): Deleted.
(WebCore::upperBoundForLengthForSubmission): Deleted.
* Source/WebCore/html/HTMLTextAreaElement.h:

Canonical link: <a href="https://commits.webkit.org/260861@main">https://commits.webkit.org/260861@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f58777c65796ba5905481dbe35fa6bf8df3e3e7a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109668 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18786 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42411 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1147 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118790 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113570 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20256 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9986 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101937 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115419 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15077 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98308 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43297 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97051 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29960 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85082 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11517 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31302 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12173 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8240 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17539 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50911 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7531 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13917 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/HTMLTextAreaElement.cpp:
```diff
@@ -57,32 +57,15 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(HTMLTextAreaElement);
-        if (text[i] == '\n')
-            count++;
+    unsigned crlfCount = 0;
+    for (unsigned i = 0; i < length; ++i) {
+        if (text[i] == '\r' && i + 1 < length && text[i + 1] == '\n')
```
Not very important, but I think this would be slightly more readable if the loop condition was changed to `i < length - 1`, and the `if` statement was simplified.

darinadler replied:
I like that idea! But if we literally did that, then we would break the zero-length string case. It needs to stay as `i + 1 < length` or some other idiom that does not underflow.

darinadler replied:
But not only is that more readable, it’s likely a tiny bit more efficient as well. Since we are only searching for CRLF, we can do it this way.

pxlcoder replied:
Good point – missed the `unsigned` 😅

cdumez committed:
Change the linebreak behavior of minlength and maxlength attributes of TEXTAREA
https://bugs.webkit.org/show_bug.cgi?id=249916
rdar://103856355

Reviewed by Darin Adler.

Align the linebreak behavior of minlength and maxlength attributes of TEXTAREA with
the specification, Blink and Gecko.

Old behavior:
 CRLF, CR, or LF is counted as 2 because we limited the submission value.

New behavior:
 CRLF, CR, or LF is counted as 1 because we limit the API value.

This is a cherry-pick of Blink:
https://chromium.googlesource.com/chromium/src.git/+/a15474ff7adff41a8956ab436ac63a7d5b7a090f%5E%21/#F3

* LayoutTests/fast/forms/textarea-maxlength-expected.txt:
* LayoutTests/fast/forms/textarea-maxlength.html:
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::computeLengthForAPIValue):
(WebCore::HTMLTextAreaElement::handleBeforeTextInsertedEvent const):
(WebCore::HTMLTextAreaElement::sanitizeUserInputValue):
(WebCore::HTMLTextAreaElement::validationMessage const):
(WebCore::HTMLTextAreaElement::tooShort const):
(WebCore::HTMLTextAreaElement::tooLong const):
(WebCore::computeLengthForSubmission): Deleted.
(WebCore::numberOfLineBreaks): Deleted.
(WebCore::upperBoundForLengthForSubmission): Deleted.
* Source/WebCore/html/HTMLTextAreaElement.h:

Canonical link: https://commits.webkit.org/260861@main
